src/main/java/build/buildfarm/instance/stub/StubInstance.java

```diff
@@ -240,23 +251,90 @@ public void write(byte[] b, int off, int len) throws IOException {
 
   @Override
   public InputStream newStreamInput(String name) {
-    throw new UnsupportedOperationException();
+    Iterator<ReadResponse> replies = bsBlockingStub
+      .get()
+      .read(ReadRequest.newBuilder().setResourceName(name).build());
+    return new InputStream() {
+      InputStream chunkIn = nullInputStream;
+      boolean closed = false;
+
+      @Override
+      public int available() throws IOException {
+        if (closed) {
+          throw new IOException("stream is closed");
+        }
+        if (chunkIn.available() == 0) {
+          advance();
+        }
+        return chunkIn.available();
+      }
+
+      @Override
+      public int read() throws IOException {
+        if (closed) {
+          throw new IOException("stream is closed");
+        }
+        byte[] b = new byte[1];
+        if (read(b) == -1) {
+          return -1;
+        }
+        return b[0];
+      }
+
+      @Override
+      public int read(byte[] b, int off, int len) throws IOException {
+        if (closed) {
+          throw new IOException("stream is closed");
+        }
+        int totalLen = 0;
+        while (len != 0) {
+          int readLen = chunkIn.read(b, off, len);
+          if (readLen == -1) {
+            advance();
+            readLen = chunkIn.read(b, off, len);
+            if (readLen == -1) {
+              return totalLen == 0 ? -1 : totalLen;
+            }
+          }
+          len -= readLen;
+          off += readLen;
+          totalLen += readLen;
+        }
+        return totalLen;
+      }
+
+      // FIXME efficient skip
+
+      @Override
+      public void close() {
+        closed = true;
+      }
+
+      private void advance() {
+        if (replies.hasNext()) {
+          chunkIn = replies.next().getData().newInput();
+        } else {
+          chunkIn = nullInputStream;
```

Comment:
This seems awkward. What if you have a chunk that happens to be zero-lengths?

Reply:
Agreed, I'm chucking the nullInputStream for ByteString.EMPTY.newInput() and iterating in advance() while hasNext and the data retrieved is empty.